### ISSUE TITLE
New Results Thread

### ISSUE THREAD

**Issue by m-orton** [m-orton/Evolutionary-Rates-Analysis-Pipeline#42]
Making a thread to post updates on results.

Chilopoda results are now up on dropbox. Alignment looked great, no gaps or issues. Chilopods only had one pairing but it was a really small dataset.

**Comment by m-orton** [comment]
Just went through and updated Mollusca, Cnidaria, Branchiopoda and Annelida server runs on dropbox with relative distance dataframes.

**Comment by m-orton** [comment]
Collembola results now up, alignment looks really good. No issues with this one.

**Comment by sadamowi** [comment]
OK great. Thank you Matt. I am starting to go through all of these results and will continue throughout this week. I am also collating the results into a results table for the manuscript. I will add "SJA" to folder names for taxa for which I have checked the final trimmed alignments and the results. Please do direct my attention towards the results any time you would like my input on any particular issue. Otherwise, I will just systematically go through the results.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, 

I posted the alignment results Im getting for Echinoderms, I wanted to get your thoughts on them. It appears as though Holothuroidea and Echinoidea both have certain sequences that have been amplified further downstream of the reference start position. In Holothuroidea there are some instances of 149-150 bp downstream of the reference sequence start position and in other other cases 169-170 downstream. In Echinoidea there are some instances of sequences being amplified 10, 20 and 40 bp downstream. 

I think this might also be a potential explanation for why some sequences end up being shorter after trimming than reference seq length.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Hi Matt,

Thank you for posting these results. Likely, various sets of primers have been used for these taxa.

At this point, I'd like to suggest that we implement our solutions/rules consistently when encountering similar issues across taxa. Here, I suggest to implement the solution that I believe you applied to Mollusca, involving applying a final length filter (after aligning and trimming against the reference sequence).

Do you agree here?

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, agreed, I will use the same length filter I used with Mollusca and should be able to post the results shortly. Also, I'm close to finishing the results for Malacostraca as well.

Best Regards,
Matt

**Comment by m-orton** [comment]
Echinoderm full results are now up, alignments look good. Actually getting significant pvalues when you count members of all classes!

**Comment by sadamowi** [comment]
Thank you for re-running this, Matt. The new echinoderm alignments look very good.

**Comment by m-orton** [comment]
Maxillopoda results now up. I had to remove two bin sequences: ACM2241 and ACM2242 that were causing a 10 bp gap in the alignment. There were also a few cases of sequences amplified further downstream so I implemented the trimming after the trimmed alignment as with Mollusca. Alignment looks very good now.

**Comment by sadamowi** [comment]
Hi Matt,

Thank you very much for preparing and posting these results. We discussed today that you will verify that the unique pairings key is working correctly. Thank you very much for checking on that. Please do let me know after you complete this check, and then I will resume my final checking of the alignments and also transfer the results to the table in the manuscript file.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, just to give a progress update on this, 

I discovered the method I'm using for creating a pairing key does not create a truly unique key in large datasets. What ends up happening is in rare cases, lineages can be incorrectly paired together by the pipeline if they possess the same key and that key is shared by multiple pairings. 

The good news is that Echinoderms, Maxillopoda, Mesostigmata, Branchipoda, Chilopoda, Collembola, Annelida and Cnidaria are all free of this pairing key duplication issue, I went and checked through the pairing keys for each pairing to make sure there were no duplicates. So the results for these groups should be good to go to the results table.

You were right about Mollusca, it had two pairings that had lineages swapped due to this problem causing high relative dist values. Malacostraca also had two pairings with this problem. Leps had a few each in NA, SA and EUR/AFR. So im figuring out a better solution for these groups and the rest of the groups yet to be done. 

The good thing is that alignments should not be affected by this so alignments that are posted can still be confirmed. I only have to rerun from section 10 onwards since that is where the problem originates from. Currently working on a solution.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Hi Matt, 

Thank you very much for checking into this issue in more detail and also letting me know which groups are good to go. Is it possible to generate unique keys by assigning sequential numbers?

That is very good news that this doesn't impact the alignments, as I know it was a challenge to get through those huge lep alignments.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, good news I think I came up with a way of creating a truly unique pairing key. I'll do my best to explain and I found this really interesting so this will be a long post. 

The problem was that when I first filter the distance matrix according to the 0.15 criteria and create the initial pairing results dataframe I get a mess of a dataframe that looks like this below:
![initialpairingresults](https://cloud.githubusercontent.com/assets/15014796/22851086/b62e2b00-efe5-11e6-81f3-81a1352a24ef.PNG)

This is why I cant assign a sequential number to each pairing to start off with because actually the lineages of each paring are not grouped together to start off with - I need a way of ordering the dataframe such that I can group lineages together. 

The issue is then to figure out how to group lineages of a pairing together such that the row and column numbers (from the distance matrix) match for each lineage. Now if I try to simply order by ingroup distance alone (which is what I was doing initially) I get duplicates and lineages are incorrectly matched together:
![ingroupdistpair](https://cloud.githubusercontent.com/assets/15014796/22851088/d263505c-efe5-11e6-8e1b-308446cd5d00.PNG)

To combat this problem I came up with a simple formula to create a pairing key that I could order by and thus group lineages together. I used ingroupdistance * (row num + col num). This seemed to work most of the time but the problem as I discovered recently is that it was possible to still have duplicates of this key so that didnt work either.

So my solution now is to use a math formula called the Cantor pairing function: http://stackoverflow.com/questions/919612/mapping-two-integers-to-one-in-a-unique-and-deterministic-way

Basically it allows you to combine two integers A and B into a unique integer C in a deterministic way such that no other combination of integers would ever give C besides A and B. It is described as:
C = 1/2(A+ B)(A + B + 1) + B

So I use this pairing function to combine the row and column number of each lineage into a key that should be unique to each lineage only so that when the dataframe is ordered by key, the lineages are always correctly grouped together!
![pairresultscantor](https://cloud.githubusercontent.com/assets/15014796/22851103/0df16a82-efe6-11e6-8084-370457208ff8.PNG)

The only way the pairing function will output the same value is if the row and col number have the same ordering (you can see they are swapped for each lineage) so to get the same ordering I use min and max as you can see above.

Hopefully that all makes sense lol and I should be able to get back on track with running through the remaining taxa.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Hi Matt,

Thank you for your work on solving this. This is indeed very interesting! I think that's you've come up with a creative solution for this issue.

What you explained makes sense to me! I suggest to verify at the end (particularly for the largest datasets) that all pairing keys are indeed unique, to ensure everything is working as it is supposed to work. As well, I suggest that we check a few pairs in the final output to make sure nothing is getting shuffled during these steps.

That's great that we should be back on track for soon generating the results for remaining groups. I have checked all of the available alignments (finaltrim alignments), and I agree that all are either very good or acceptable for analysis. I have transferred the results to the results table in the manuscript draft file, for those taxa you indicated were good to go.

Best wishes,
Sally

**Comment by sadamowi** [comment]
Hi Matt,

Would you please let Jacqueline know when the revised section of code mentioned above for creating the pairing key is posted to Github? I suggest it would be a good idea for Jacqueline also to check over the revised code. Thank you Matt and Jacqueline.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally and Jacqueline, 

The revised section of code for pairing keys is now up on both versions of the script. Lines 964-992 in the small taxa version of the script and lines 1041-1069 on the large taxa version.

Sally - I've gone through and run through both Malacostraca and Mollusca again with the revised code to ensure that pairings are being generated correctly and retained the pairing key in the pairing results so that it can be double checked. You will now notice there are no high relative distances in any of these groups also. Additionally, I now order pairing results by pairing key to ensure that ingroup lineages are correctly grouped together. 

This should not have any effect on alignments (since the alignments are performed at an earlier step in the code) however it will have an effect on the final results for each group considering that lineages were not being correctly grouped together in these groups previously.

Anyways I think I've managed to fix the problem and am now working on finishing up the other taxa to be done. I'm thinking that what I will do is try and get all the smaller taxa done first including all of Chordata and then once I'm finished those I will work on revising Leps and finishing the other insect orders. Let me know if you think this is a good way of doing things.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Hi Matt,

That sounds great. Thank you again for finding this solution. And, that would be a good idea if Jacqueline would check this new section of code as well.

I think that is a good plan to work first on completing the smaller groups. Once Chordata is finished, I can work on certain sections of the draft that relate to some groups within Chordata, and Jacqueline can also run them with her phylo pipeline for comparison. While we are doing those things, you could be completing the remaining insect orders.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, 

Results are now up for Pycnogonida, Ostracoda, Perciformes and Cypriniformes! Alignments were good for all groups. No significant values but Perciformes did have a near significant wilcoxon pvalue which I thought was interesting. Pairing results for all of these groups also have the pairing key in the results summary to ensure pairings are being generated correctly. Rest of Chordata to follow tomorrow!

Best Regards,
Matt

**Comment by m-orton** [comment]
Alright so Chordata is now up in its entirety. I was able to do the remaining classes in one runthrough: Reptilia, Aves, Elasmobranchii, Mammalia and Amphibia. Fortunately, the alignments all look really good! Reptilia sorry to say didnt have any pairings,  Amphibia had one, Mammals and Elasmobanchii had a dozen or so each but Aves pretty much had the vast majority. No significant values to report. 

As a side note: the only reason I couldn't include fishes in a single Chordata runthrough was because I had to separate Perciformes and Cypriniformes due to their separate reference sequences.

Also, from now on, I think what I will do is compress the section20 files when I upload to dropbox to keep upload times and file sizes minimal.

Best Regards,
Matt

**Comment by m-orton** [comment]
Spider results are now up as well!

With Araneae, I had to eliminate two bins: ACS2027 which had an 8 bp insertion and ACL2573 which had a 1 bp insertion.  I also had to incorporate the Mollusca trimming code after the trimmed alignment since there were a few sequences shorter than reference length - most likely due to primers amplifying a different region. But the alignment looks good now after making these changes. Very good number of pairings for Araneae also!

I also checked the pairing key and found no duplicates which is a good confirmation that the new code is working since Araneae is such a large group.

Also, I made a few small edits to the code today on Github - edited a small part of the pseudoreplicate code, made a few changes to comments and fixed a minor error I encountered with the largetaxa version of the script. The details can be found in the commit comments.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Dear Matt,

Thank you for these posts and progress. I look forward to going through the alignments and further filling in the Results table within the next 1-2 days. I should have mentioned earlier that I had designated additional reference sequences for a couple of the larger orders of fish and arachnids just in case you wished to run them separately due to their size. That's great we are on the homestretch with the V1 full results.

Best wishes,
Sally

**Comment by m-orton** [comment]
Coleoptera results are now up. 

Coleoptera had 4 bins which I manually removed for the alignment. ACZ1158 had a large 22 bp insertion, ACZ1156, ACO5660 and ACM2866 all had small 1 bp insertions. Over 900 pairings though! 
As with Araneae I implemented the Mollusca sequence trimming code after the trimmed alignment to remove shorter sequence since there were a couple in trimmed alignment.

I think the pvalues are also pretty definitive since we are not getting significance even with such a large group. Also the pairingkey solution I implemented seems to be working well.

Best Regards,
Matt

**Comment by m-orton** [comment]
Posting my timeline for the remaining groups:
Fixing Lepidoptera with pairingkey solution: 1-2 days
Diptera (dividing by region): 2-3 days
Hymenoptera (dividing by region): 2-3 days

Thats it, we are finished!

**Comment by sadamowi** [comment]
Hi Matt,

That's great we are on the home stretch!

Just to be sure there was no misunderstanding about the extra REF seq I designated for Arachnida (because Araneae was a large order), did you run the remaining Arachnida, aside from Mesostigmata and Araneae?

Thank you very much.

Cheers,
Sally

**Comment by m-orton** [comment]
Hi Sally, sorry I havent run the rest of Arachnida, could you resend me the ref seq for Arachnida? Couldnt find it on dropbox. I will then get the rest of Arachnida done today.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Hi Matt,

I think the Aranaea REF sequence could be used. I gave two ref seqs for that class in case you wanted to split up the class, as there were a couple of larger groups. But perhaps it would have run all together. Thank you!

Cheers,
Sally

**Comment by m-orton** [comment]
Hi Sally, 

I did a runthrough for the rest of Arachnida and the results are now up. I think it was good to separate Araneae since it did end up being a really large group and we already have the full results for Mesostigmata. The Arachnida runthrough went well overall, there were three bins I had to remove AC05111, AAN9613 and ACK5180. All of these had 1 bp insertions, as with Araneae I implemented the Mollusca sequence trimming code after the trimmed alignment to remove shorter sequences.

Best Regards,
Matt

**Comment by sadamowi** [comment]
Excellent. Thank you Matt.

**Comment by m-orton** [comment]
The new Lep AUS results are up now with no duplicates of pairing key. I also decided to fix the 1 bp insertions in the alignments by removing 8 bins: AAD6260, AAE6214, AAE6215, AAD2088. AAF4403, AAE9708, AAI1270 and AAI1271. Alignment is looking great now.

**Comment by sadamowi** [comment]
Super! I will be going through these new results today and tomorrow.

PS. Just to clarify ... Will there be new lep results for the other geographic regions as well at some point? Or, did the pairing issue only influence AUS?

**Comment by sadamowi** [comment]
Hi again Matt,

I was wondering if you have run the remaining orders in class Actinopterygii? I see results for the orders Perciformes and Cypriniformes. In this case as well, I had provided two reference sequences for this class because a couple of the orders are large. However, I am happy to have orders be run together that work well. Perhaps all other orders together except those two large ones?

Thank you for checking.

Best wishes,
Sally

**Comment by sadamowi** [comment]
Hi Matt,

I have looked through all the new results and added them to the results table in the manuscript draft file. Everything looks good to me! 

Best wishes,
Sally

**Comment by m-orton** [comment]
Awesome, thanks Sally. No I actually haven't run the rest of Actinopterygii yet, I think I misunderstood the reference seq file since I've only been running the taxa that have a ref sequence for them in that file. Sorry about that, I'll get through the rest the fish orders today.

**Comment by sadamowi** [comment]
No worries Matt. I should have explained the chosen ref seqs more clearly. Not a big issue. Thank you for running the rest of that fish class.

Best wishes,
Sally

**Comment by m-orton** [comment]
Rest of fishes are up! Really pairing rich group with around 500 pairings. Had to remove 3 bins with small insertions: ACG9172, ADC2507 and ACS9986 but the alignment is looking good now.

**Comment by jmay29** [comment]
Hello!! I just did a check of the pairing keys for a few orders and I did not get any keys that linked to more than 2 (they all appear to be unique)! I checked on both the small and large taxa pipelines.

**Comment by sadamowi** [comment]
Thank you for checking this new code, Jacqueline.

**Comment by m-orton** [comment]
Thanks Jacqueline! Just posted Lep SA, results looks good, removed AAY0441 due to a 1 bp insertion.

**Comment by sadamowi** [comment]
Great - on the homestretch!

**Comment by sadamowi** [comment]
Hi Matt,

This is just a friendly remind that for Insecta, I only selected reference sequences for the largest orders - the ones that we had discussed would need to be be run separately. There are lots of other orders within Insecta. Perhaps that those be run all together? If you need any additional reference sequences, please let me know.

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, no problem, I think it should be ok to run the rest of Insecta as one group. Which of the ref sequences for Insecta could I use, could I just use the Lep ref sequence or Coleoptera ref sequence?

**Comment by sadamowi** [comment]
Thank you Matt. From a theoretical perspective, it would not matter which reference you use, as all of these orders currently having a REF seq are fairly related to one another and also equally phylogenetically distant from the majority of the other insect orders. However, I would advise against choosing Hymenoptera, as rates of molecular evolution are high in some members of that order. You wouldn't want to pick a reference from an insect group with a "weird" rate of molecular evolution. I suggest that Lepidoptera would be a good choice. Primers designed using Lepidoptera sequences work well across a large range of insects. Therefore, I would expect these to have relatively "typical" sequences for insects.

Best wishes,
Sally

**Comment by m-orton** [comment]
Ok sounds good, I'll go with the Lep ref sequence then.

**Comment by m-orton** [comment]
Lep NA is now up, had to remove 4 bins to get the alignment looking good, AAF0261, AAC0205, AAG1173 all had 1 bp insertions while AAK7666 had a 2 bp insertion.

**Comment by sadamowi** [comment]
OK great - thanks Matt.

**Comment by m-orton** [comment]
Lep EUR/AFR is now up so Lep is now officially done! Only had to remove 1 bin AAE5820, this bin had quite a large insertion of 59 bp.

**Comment by sadamowi** [comment]
hooray!

**Comment by m-orton** [comment]
Lots of groups posted!

Insecta (not including Lep, Col, Dip and Hym orders) has been posted. There were a number of 1 bp insertions which I removed: ACI7875,ACA7403, ACT9180, ACN6142, ACW8226, ABY1171. There were also a few real indels (or what appeared to be real indels) but alignment looks good.

I also posted Diptera AUS, SA and EURAFR. No bins that I had to remove with these. EURAFR had a few 3 bp indels which I left in and I also had to use the trimming code as well. EURAFR also has significant values which is interesting for a group with over 450 pairings. 

Just have to do NA and then on to the last group Hymenoptera.

**Comment by sadamowi** [comment]
exciting! Great news!

**Comment by m-orton** [comment]
Alright so Hymenoptera is now finished and up on dropbox. 

A few of the regions had 3 bp indels so I left those in the alignments.

SA and AUS for Hymenoptera went smoothly without having to remove any bins. 
EURAFR I had to remove 1 bin ACU9389 for a 1 bp insertion. NA I had to remove several bins for 1 bp insertions: ACI5609, ACJ5404, ABV2664. ACT0024, ACV9606, ACL7936, ACX6770 and ACC4201. NA also had pvalues of 0.050 and 0.038 for binomial and wilcoxon.

Sally, could you take a look at the alignments for Diptera NA I posted when you get a chance? I assume I should remove the 1 and 2 bp insertions but there is a 9 bp insertion in particular I wanted to get your thoughts on before continuing with this group. 

Also, should I be doing a run of the classes that remain for Chordata? There are a few small classes in Chordata we haven't run although they are small enough that they might not yield any pairings.

**Comment by sadamowi** [comment]
Hi Matt,

This is great! I will do my best to look very soon at the alignments you mention.

At the class level, I selected reference sequences for those classes that generated pairs in your preliminary analysis. That list of preliminary pairs is saved as a tab in the reference sequence Excel file. So, I believe we should already be good in terms of running Chordata classes.

Do you agree?

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, thanks for checking the alignment.

Agreed, looking at the preliminary pairs I think we should be good with the Chordata classes we have run.

**Comment by sadamowi** [comment]
Hi Matt,

Thank you very much for reorganizing the files in Dropbox. This is very helpful. I realized I had one file open, but I since moved that file too. We ended up with a lot of runs, and so I am very glad that we decided to go with the two major pipelines (large taxa, small taxa) on github. I believe some of the runs were due to the slight misunderstanding over the REF sequences. For example, potentially in the future (e.g. if we need to rerun things due to reviewer comments), some of the smaller arthropod classes could be run all together and/or perhaps all of Actinopterygii could be run together. Anyhow, we can consider overall how to make things more efficient for any future needed analyses. I will now look at the specific Diptera alignments mentioned above and post again shortly ...

**Comment by sadamowi** [comment]
Hi again Matt,

I have had a look at Diptera NA. I suggest to remove:

ABY1630, ACB0007, ACB9011, ACC9152, ACD1624, ACD9327, ACG4331, ACI6633, ACJ3130, ACJ8384, ACX9549, ACY9512 - Reason - Total of 12 bp worth of insertions about 130 bp in from the beginning of sequence.  As well, I think that these sequences may be the cause of the gaps in other sequences flanking these areas.

These sequences are highly unusual for Diptera. However, this is very interesting. These could be real biological variants. I built a tree for just these sequences, and they are close relatives. Also, I blasted one and it is Cecidomyiidae, a family with an unusual breeding system (haplodiploidy). There does seem to be a very interesting pattern of highly gapped sequenced being associated with unusual breeding systems (as was the case with the unusual mite family that we detected as well). Others have pointed out associations between breeding systems and evolutionary rates (which appears possibly to include rate of indels as well). However, I think exploring this further at this time is beyond the scope of the latitude paper.

I agree with you about leaving out sequences causing 1-2 bp indels (if you have the patience for that). We previously decided to leave in such cases when the errors occurred near the end of the sequence, as the most likely explanation was base calling error as the signal weakened near the end of the sequence. Such cases of occasional indels would not influence the results because of our usage of pairwise deletion of missing data. However, sequences that contain many such cases are more problematic, because that would be a signature of a nuclear pseudogene.

After these deletions, how does the Diptera NA alignment look?

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally,

Thats interesting, I didnt know about haplodiploidy. I just got through the alignments for Diptera NA removing the bins you mentioned and also removing several other bins with 1-2 bp indels. I posted the alignments I ran today (adding Mar6 to the title of the alignment). I think the alignment looks much improved now. 

Something that might be useful to you and Jacqueline, I think I found an easy to identify bins with small 1-2 bp insertions that are hard to find in large alignments. I open up the alignment in MEGA, then copy the column with the insertion over to a text editor and then use the find function in the text editor to search for nucleotides. Seems to work well.

Best Regards,
Matt

**Comment by m-orton** [comment]
Happy to report Diptera NA is up and all results are now posted, hooray!

Also, I'm going to update shortly the weirdbins excel file with the rest of the problematic bins I have found in Insecta.

**Comment by sadamowi** [comment]
hooray! And, congratulations on this important milestone for the project!

**Comment by sadamowi** [comment]
Hi Matt,

For Perciformes, here is the list of pseudoreplicates:

("63", "1")
("67", "2")
("15", "11")
("46", "12")
("42", "17")
("46", "24")
("39", "56", "26")
("45", "29")
("33", "32")

I notice that in one case, pseudoreplicates are grouped together into a trio. In another case, pair 46 appears in two different pseudoreplicate pairs. I wanted to let you know about this in case you think the pseudoreplicate section could need tweaking for the pipeline version.

For the Excel calculations, I am creating a new trio.

Best wishes,
Sally

**Comment by sadamowi** [comment]
Hi Matt,

I have noticed that a few pairs have generated very extreme branch length ratios (even one case of -37). In every case that I have seen so far, this related to very closely related BINS (<2% divergence). I confirmed these very genetically similar BINs by looking at the BIN database on BOLD. So, I think the pairing function is working correctly. I knew that BINs could be formed that are closer than 2%, based upon the patterns of genetic variability, but I am a little bit surprised to see multiple cases, including BINS with close to just 1% divergence. I am wondering if perhaps the BIN threshold has been tuned for certain taxonomic groups (to give groupings closer to species). I will check with BOLD so that we have a clear understanding of this issue.

This issue of very small genetic distances being associated with highly variable branch length ratios has been discussed before in the literature. A lot of error in these estimates is expected at very small genetic distances, for example if one lineage has no or few substitutions since the divergence event.

I suggest that we consider omitting pairs with <2% divergence. What do you think? I can check for this with the 5 phyla I am working with now (all but Arthropoda). What do you think?

Best wishes,
Sally

**Comment by m-orton** [comment]
Hi Sally, yes I do agree with the <2% divergence rule. I think that would be a good idea to implement to avoid cases of really large relative outgroup distances. Would you want me to also implement this for Arthropods?

For the pseudoreplicates, I think for now we will need to check the results for duplicates but hopefully in a V2 I can fix this issue.

Best Regards,
Matt

**Comment by sadamowi** [comment]
OK thanks Matt. I'm glad you agree. Yes, please do implement the rule that the IG distance must be at least 2%. I am checking for this in the 5 other phyla.

Best wishes,
Sally

**Comment by sadamowi** [comment]
If a pair with a very small IG distance falls in a pseudoreplicate, I am omitting the pair with the small IG distance and retaining the other member of the pseudoreplicate grouping.

**Comment by m-orton** [comment]
Hi Sally, I finished coding the relative branch length, 2% and pseudoreplicate sections. I'm just wondering for the binomial test if I can use the number of positive values as successes from relative branch length as opposed to relative distance? Its just makes things easier from a coding perspective so I dont have to do both relative dist and relative branch length calculations.

I realize the values are different but the signage would be the same right for relative dist vs relative branch length?

**Comment by m-orton** [comment]
Also, I just tested Perciformes to test the relative branch length code and I'm getting the same values as you so I'll post an excel in Perciformes with my result so you can double check as well.

**Comment by sadamowi** [comment]
OK great! Thanks Matt. I was also thinking we should do a test to make sure our approaches agree. So, that's good to hear.

Best wishes,
Sally

**Comment by sadamowi** [comment]
Hi Matt,

I've had a look at your file and cross-checked a subset of the data. We got exactly the same values - good! It would appear that this file was for checking the branch length ratios, prior to accounting for pseudoreplicates and prior to omitted pairs with low ingroup divergence?

I'd like to suggest a few minor additions:

-I suggest to output the estimated branch lengths for each paired lineage (in addition to the signed branch length ratio). These raw data may be useful for us.

-in the pval dataframe (or another dataframe if easier), I suggest also to export a count of the positives and negatives and the median signed branch length ratio. These values are useful in addition to the p-values.

-I am also further considering two ways of calculating the ratios. The method I am more familiar with is using signed ratios. Another method I have seen in the literature is always going one group over another (e.g. always low-latitude divided by high-latitude estimated branch length). Do you have thoughts on this? I am reading about this further. This would be an easy calculation to add, as the formula is very simple compared to what we've already done. i.e. This would be in addition to the signed ratios, and we would report both. I think these two different methods are also contributing to the variability in effect sizes reported in the literature. What do you think about trying both? We could the median values from both calculations to Table 1, and these values would be helpful for the discussion.

Best wishes,
Sally

**Comment by sadamowi** [comment]
PS. An example of the other type of ratio is Gillman et al. 2009 (e.g. see supplementary table). There are no negative values. "Opposite direction" values would fall between zero and 1. Ratios are always calculated in order of the habitat (e.g. lower latitude over higher latitude), rather than larger over smaller (plus a sign). So, I'm thinking it could be helpful for us to do the ratios both ways. Thoughts?

**Comment by sadamowi** [comment]
PPS. Another idea is only to do the alternative branch length ratio method for select taxonomic groups (such as fish), to enable direct comparison with other studies in the discussion. I can readily do this just for those select groups if the data are outputted so as to include the branch lengths for each sister (in addition to the signed branch length ratio). This could be a better approach in terms of keeping the manuscript more streamlined. As well, I think the signed ratios are easier to interpret. e.g. 1.05 is 5% higher rate at low latitude and -1.05 is 5% higher rate at high latitude. Thanks for any thoughts on this.

**Comment by m-orton** [comment]
Hi Sally, the test with Perciformes was before pseudoreplicates and low ingroup divergence just to test I was getting the same result as you (which I did, yay!). I have also incorporated those changes into the code as well and I updated the code on Github. 

The minor additions you mentioned should be easy to do so I will incorporate those into the pairingresultssummary dataframe (estimated branch lengths) and pval dataframes (counts and median signed ratios). 

I think in regards to using two different versions of the ratio's, if its easy to implement (which it seems like its just a simple calculation) then I think it would be a good idea to report both. I think it would only strengthen our results to do both. 

I'm just wondering if we would do two different wilcoxon tests though for each set of ratio's or would it only be the signed ratio's that get inputted to the wilcoxon test? Also, can I just use the number of positive signed branch length ratios as the number of successes for the binomial? (just checking to be sure)

**Comment by sadamowi** [comment]
Hi Matt,

Thanks for your reply on this and for implementing those suggested edits to output more information for us.

The binomial test would only need to be performed once. For the signed ratios, it would be the number of positives compared to the total count. For the lower/higher latitude branch length ratio, it would be the number of values greater than 1 vs. the total, and this would  yield the same result as for the other ratio.

For the Wilcoxon test, the lower/higher latitude branch length ratio would need to be run against a null expectation of mu=1 (rather than mu=0 - zero is for the signed ratios). The results can be somewhat different as the rankings of values can be different.

Note that I still have the manuscript file open. I should be able to pass that back to you later today.

Best wishes,
Sally

**Comment by m-orton** [comment]
Thanks for clarifying Sally, I will make the modifications to the code today.

Best Regards,
Matt

**Comment by m-orton** [comment]
Hi Sally,

Just posted the new results for Actinopterygii, old fish results were moved to old results folder. 
There were two BINs I thought I should mention with really large insertions of 48 bp and 44 bp:
ADA2828, ADC2043 so those were removed, alignment looks good now.

Actinopterygii was run with all of the new code additions which I just finished so you can see how I am now generating results:
-2% divergence code
-Pairing results summary shows both types of ratios and estimated branch lengths
-Modified Sections 16-19 to incorporate the calculation of alternative branch length ratios - lower lat / higher lat 
-Pseudoreplicate averaging will now average these new branch length ratios in addition to the signed ratios
-Wilcoxon tests will now be performed on these new ratios (low / high) with mu = 1. Signed still uses 0
-Pvalue dataframe will include pvalues for both wilcoxon tests, number of positive and negative signed ratios and median values for both signed and low lat / high lat ratios

There are slighlty fewer pairs than before but I am using the 2% code so that reduces number of pairs...and Im getting significant values now!
0.009543 for binomial, 0.003536 for wilcoxon (signed), 0.000281 for wilcoxon (low/high) - also seems to corroborate the fish paper findings (despite the papers flaws we talked about).
So I think we made the right decision to switch to relative branch length.

Also I'm calling the new estimated branch lengths (low/high) "altRelativeBranchLength" for simplicity. 
The branch lengths are identical between methods when the value is positive for signed ratio's but different when negative since the new method cant have negatives.
Averaged pseudoreplicate branch lengths are different because of the method we use for averaging signed ratio's
(adding or subtracting 1 to avoid decimals between 0 and 1). Wilcoxon results differ I think because of this but median ratios between the two methods also differ slightly.

Lastly, got lazy with making excels from the results so I also added in a write excel 
command to the code that will write all results directly to a single excel file and the name will reflect 
the class and date when the analysis was run. Each results dataframe will also appear in a different spreadsheet of the same file. Added this command at the end of the statistics section. This should allow me to rapidly go through and regenerate results for Arthropods!

Best Regards,
Matt

**Comment by m-orton** [comment]
This issue was moved to jmay29/lat-project#4
